### PR TITLE
Service discovery fixes

### PR DESCRIPTION
**Pull request by fabxc**
Removes closing of update channels and fixes all missing respects of the context.
Changes across various, mostly recent, SDs. Almost all of these are potential deadlocks "only" manifesting as goroutine leaks.
This needs careful review so we don't break anything further.

@brian-brazil the second commit consolidates the Zookeeper SDs. As far as I could see those were just distinguished by their parsing function.
Those no longer send the initial set of all targets, i.e. could have a bit of flapping now on config reload. However, what was there before was realistically not working anyway as it just immediately sent whatever it got between constructing the SD and calling `Run()`.
If someone cares enough they are free to implement a proper initial walk later.

@juliusv

### CODE REVIEW

File: retrieval/discovery/zookeeper.go [hunk 1]
```diff
@@ -30,111 +29,112 @@ import (
...
+
+	for {
+		select {
+		case <-ctx.Done():
+		case event := <-sd.updates:
+			tg := &config.TargetGroup{
+				Source: event.Path,
+			}
+			if event.Data != nil {
+				labelSet, err := parseNerveMember(*event.Data, event.Path)
```
Comment (fabxc) [inline retrieval/discovery/zookeeper.go]: outch. fixed

File: retrieval/discovery/zookeeper.go [hunk 2]
```diff
@@ -30,111 +29,112 @@ import (
...
+	return NewZookeeperDiscovery(conf.Servers, time.Duration(conf.Timeout), conf.Paths, parseServersetMember)
+}
+
+// NewZookeeperDiscovery returns a new discovery along Zookeeper parses with
+// the given parse function.
+func NewZookeeperDiscovery(
+	srvs []string,
+	timeout time.Duration,
+	paths []string,
+	pf func(data []byte, path string) (model.LabelSet, error),
```
Comment (fabxc) [inline retrieval/discovery/zookeeper.go]: fixed

**Comment by fabxc** [comment]
> This was a consequence of the SD API at the time making this impossible to implement correctly. It's since changed so that's no longer the case.

Yes, indeed. The API didn't support sending target group slices but the case that was added here once it did, was effectively not doing anything. I think it was myself adding it back then.

**Comment by fabxc** [comment]
@juliusv @brian-brazil does this look good for merge?

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**